### ISSUE TITLE
Allow sniffs to be selectively included

### ISSUE THREAD

**Issue by PatrickRose** [squizlabs/PHP_CodeSniffer#656]
Hi,

I have a few Sniffs that should only apply to a few directories, so I can't use an exclude. Is there way I can tell PHPCS to whitelist only a certain ruleset (failing that, how should I build an exclude)?

**Comment by gsherwood** [comment]
The exclude rules can be regular expressions. So you'd need to write a regex that only matches the directories that you want, and make that negative (probably using a negative lookahead).

There isn't an include tag in the ruleset yet, but it is something that would be very good to add.

**Comment by gsherwood** [comment]
I've added this feature to ruleset.xml files in 3.0.

The example below tells PHPCS to only apply the ElseIfDeclaration sniff to files in the `tests` and `templates` directories.
```xml
<rule ref="PSR2.ControlStructures.ElseIfDeclaration">
    <include-pattern>*/tests/*</include-pattern>
    <include-pattern>*/templates/*</include-pattern>
</rule>
```
As with exclude patterns, you can have as many include patterns as you want, and they support regular expressions in the same way.

**Comment by white-gecko** [comment]
Is it also possible to combine `exclude-` and `include-pattern` in a way?

**Comment by gsherwood** [comment]
@white-gecko Using both exclude-pattern and include-pattern doesn't make any sense. Excluding assumes you are running the sniff for all files expect a few. Include assumes you are only running the sniff for a few files.

So using an include-pattern is there to save you having to write a huge list of exclude patterns.

What are you trying to achieve that would require both?

**Comment by white-gecko** [comment]
My use case is, that I'm generally checking the complete code base, but excluding the extensions/plug-in directory, except for some core-extensions, which should as well be checked.
(But I think I will be able to do this with a regular expression in an `exclude-pattern` as well.)

**Comment by white-gecko** [comment]
For the reference:

    <exclude-pattern>extensions\/(?!account\/|application\/|auth\/).*</exclude-pattern>

does what I wanted to achieve. (actually my regex is longer, since it has more directories in it)

Maybe this was more kind of a support question, sorry and thank you.
